### PR TITLE
kimi k3 for titanrl

### PR DESCRIPTION
**Pull request by liangel-02**
as title, add titanrl support for k3. for batch invariant mode, use chunked replay kernel (cudagraph friendly and paging aware) we match decode to prefill instead of prefill to decode since its faster

batch invariant bitwise parity passes 
verified 0 logprob diffs
Screenshot 2026-09-14 at 3 12 41 PM

Screenshot 2026-09-18 at 5 50 52 PM

### CODE REVIEW

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -450,11 +450,6 @@ def observe(
             return
 
         with spmd.no_typecheck(), torch.no_grad():
-            if padding_mask_T is not None:
-                valid_mask_T = ~padding_mask_T
```
Comment (liangel-02) [inline torchtitan/models/common/moe.py:454]: this was breaking cudagraphs for k3

**Comment by felipemello1** [comment]
This is awesome! thank you! Can you please run it for DAPO math for at least 50 steps and show improvement in the rewards? i guess you cannot do it in single node, right? we have the mast launcher. I would recommend something around 2 or 3 generators for 1 trainer node ratio. If you can run for 150 steps, thats better. It should be straight forward for claude/codex to adapt the dapo math 8k config we have for qwen 3.5. I would prefer to have this as our default rather than alphabetsort, if you dont mind, and publish the reward curve in some .md.

File: tests/integration_tests/rl.py [hunk 2]
```diff
@@ -286,7 +316,10 @@ def run_single_test(
             f"RL integration test: {test_flavor.test_descr}, command: {cmd} ====="
         )
 
-        result = subprocess.run(cmd, text=True, shell=True)
+        env = os.environ.copy()
+        if test_name == "rl_grpo_kimi_k3_debug_batch_invariant":
+            env["VLLM_SSM_CONV_STATE_LAYOUT"] = "SD"
```
Comment (tianyu-l) [inline tests/integration_tests/rl.py:321]: can this not be set at places like https://github.com/pytorch/torchtitan/blob/main/torchtitan/rl/generator.py#L858

File: tests/integration_tests/rl.py [hunk 3]
```diff
@@ -233,6 +235,34 @@ def build_rl_test_list() -> list[OverrideDefinitions]:
...
+                    "--metrics.no-enable-wandb",
+                ],
+            ],
+            "RL GRPO Kimi K3 hybrid KDA batch-invariant",
+            "rl_grpo_kimi_k3_debug_batch_invariant",
+            ngpu=4,
+            disabled=not (
+                torch.cuda.is_available()
+                and torch.version.hip is None
+                and torch.cuda.get_device_capability() in ((10, 0), (10, 3))
```
Comment (tianyu-l) [inline tests/integration_tests/rl.py:263]: this should run in B200 CI machine -- could you work with @wwwjn on that

File: torchtitan/models/kimi_k3/kda.py [hunk 4]
```diff
@@ -160,10 +180,12 @@ def forward(
         conv_output_1TC = causal_conv1d(
             mixed_qkv_1TC,
             conv_weight_C1W[:, 0],
-            activation="silu",
+            activation=None if is_in_batch_invariant_mode() else "silu",
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/kda.py:183]: is this because activation_fn fusion breaks batch invariance, so that we have to apply from outside?

File: torchtitan/models/kimi_k3/model.py [hunk 5]
```diff
@@ -164,9 +165,17 @@ def _apply_attention_residual(
     variance = values_float.pow(2).mean(dim=-1, keepdim=True)
     keys_TND = values_float * torch.rsqrt(variance + norm.eps)
     score_weight_D = norm.weight.float() * projection.weight.squeeze(0).float()
-    scores_TN = (keys_TND * score_weight_D).sum(dim=-1)
-    probs_T1N = torch.softmax(scores_TN, dim=-1).unsqueeze(1)
-    output_TD = torch.matmul(probs_T1N, values_float).squeeze(1)
+    if is_in_batch_invariant_mode():
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/model.py:168]: I feel the proper way is to patch the pytorch ops in https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/utils.py#L288, instead of putting if-else everywhere

File: torchtitan/models/kimi_k3/kda.py [hunk 6]
```diff
@@ -30,6 +31,19 @@
 # W = convolution kernel width.
 
 
+def _batch_invariant_silu(x: torch.Tensor) -> torch.Tensor:
+    x_float = x.float()
+    return (x_float * torch.sigmoid(x_float)).to(x.dtype)
+
+
+def _batch_invariant_l2norm(x_1THK: torch.Tensor) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/kda.py:39]: similar

File: torchtitan/rl/model/kda_attention.py
Comment (tianyu-l) [inline torchtitan/rl/model/kda_attention.py:1]: rename to kda.py to be consistent with gdn.py

File: torchtitan/rl/model/kda_attention.py [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+def _batch_invariant_silu(x: torch.Tensor) -> torch.Tensor:
+    x_float = x.float()
+    return (x_float * torch.sigmoid(x_float)).to(x.dtype)
+
+
+def _batch_invariant_l2norm(x_1THK: torch.Tensor) -> torch.Tensor:
+    x_float = x_1THK.float()
+    return (
+        x_float * torch.rsqrt((x_float * x_float).sum(dim=-1, keepdim=True) + 1e-6)
+    ).to(x_1THK.dtype)
```
Comment (tianyu-l) [inline torchtitan/rl/model/kda_attention.py:47]: let's not define them everywhere

File: torchtitan/rl/model/attention.py [hunk 8]
```diff
@@ -251,6 +271,22 @@ def forward(
         return output[:num_actual_tokens]
 
 
+class PyTorchVarlenInnerAttentionDiffKVImpl(PyTorchVarlenInnerAttentionImpl):
```
Comment (tianyu-l) [inline torchtitan/rl/model/attention.py:274]:
```suggestion
class TorchTitanVarlenInnerAttentionDiffKVImpl(TorchTitanVarlenInnerAttentionImpl):
```
following recent change of naming convention https://github.com/pytorch/torchtitan/blob/main/torchtitan/rl/model/attention.py#L38

similar for other classes in this file

File: torchtitan/rl/model/vllm_wrapper.py [hunk 9]
```diff
@@ -377,6 +394,8 @@ class _InferenceConfig:
             skip_dp=True,
         )
 
+        self._state_dict_layouts = self.get_state_dict_layouts()
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_wrapper.py:397]: why do we need this?

File: torchtitan/rl/model/vllm_wrapper.py [hunk 10]
```diff
@@ -546,7 +558,20 @@ def _maybe_initial_load_weights(self) -> None:
...
+            *self.model.named_buffers(),
+        ):
+            if not spmd.has_local_type(state):
+                continue
+            local_type = spmd.get_local_type(state)
+            partition_spec = spmd.get_partition_spec(state)
+            if partition_spec is None and all(
+                axis_type is spmd.R for axis_type in local_type.values()
+            ):
+                layouts[state_name] = SpmdType({})
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_wrapper.py:574]: what are we doing here?

File: torchtitan/rl/model/vllm_registry.py [hunk 11]
```diff
@@ -192,6 +215,97 @@ def get_state_copy_funcs(cls, mamba_types):
     model_cls.get_mamba_state_copy_funcs = classmethod(get_state_copy_funcs)
 
 
+def _configure_kda_hybrid_model(model_cls: type, model_spec: ModelSpec) -> None:
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_registry.py:218]: Let's put this under rl/model/kda.py and rename to a public function, and put the counterpart under gdn.py
```suggestion
def maybe_configure_kda_hybrid_model(model_cls: type, model_spec: ModelSpec) -> None:
```

File: torchtitan/rl/model/vllm_registry.py [hunk 12]
```diff
@@ -45,6 +47,24 @@
 # Selects the experiment-owned runner that pads tokens for dense and expert SP.
 TORCHTITAN_WORKER_CLS = "torchtitan.rl.model.vllm_worker.TorchTitanGPUWorker"
 
+_KDA_CHUNK_SIZE = 64
+
+
+def _attention_dimensions(
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_registry.py:53]: maybe put under model/attention.py and rename to

```suggestion
def get_attention_dimensions(
```

File: torchtitan/rl/model/vllm_registry.py [hunk 13]
```diff
@@ -45,6 +47,24 @@
 # Selects the experiment-owned runner that pads tokens for dense and expert SP.
 TORCHTITAN_WORKER_CLS = "torchtitan.rl.model.vllm_worker.TorchTitanGPUWorker"
 
+_KDA_CHUNK_SIZE = 64
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_registry.py:50]: defined in kda.py

File: torchtitan/rl/model/kda_attention.py [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+from torchtitan.protocols.module import Module
+from vllm.config import get_current_vllm_config
+from vllm.forward_context import get_forward_context
+from vllm.model_executor.layers.mamba.abstract import MambaBase
+from vllm.model_executor.layers.mamba.mamba_utils import (
+    is_conv_state_dim_first,
+    MambaStateDtypeCalculator,
+    MambaStateShapeCalculator,
+)
+from vllm.v1.attention.backends.gdn_attn import GDNAttentionMetadata
```
Comment (tianyu-l) [inline torchtitan/rl/model/kda_attention.py:28]: @acisseJZhong we can call ours `xxxAttentionMetadata` as well

File: torchtitan/rl/model/kda_attention.py
Comment (tianyu-l) [reply torchtitan/rl/model/kda_attention.py:1]: For GDN, @drisspg introduced this file https://github.com/pytorch/torchtitan/blob/main/torchtitan/rl/model/gdn_backend.py

For KDA do we not need to define similar things?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/rl/model/kda_attention.py
Comment (drisspg) [reply torchtitan/rl/model/kda_attention.py:1]: we should, make sure are using cudagraphs in all the places. gdn_backend is likely exactly applicable to kda so we shoudl probably just generalize it

File: torchtitan/rl/model/linear_backend.py
Comment (tianyu-l) [inline torchtitan/rl/model/linear_backend.py:1]: linear_attention_backend.py to not confuse with linear matmul

File: torchtitan/distributed/utils.py
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:1]: given we have more changes, let's extract related code into distributed/batch_invariant.py

**Review by tianyu-l — COMMENTED** [review]
looks mostly good to me, thank you

@wwwjn could you prioritize setting up the rl CI structure? or give suggestion on how to proceed if you don't plan to prioritize this?

**Comment by tianyu-l** [comment]
@claude please review the changes in moe.py

File: tests/assets/losses/rl_grpo_cuda.txt [hunk 15]
```diff
@@ -1,10 +1,10 @@
...
-2 0.0009377798996865749
-3 0.0003434247337281704
-4 -0.0010341869201511145
-5 0.000811806763522327
-6 0.0004955545882694423
-7 0.00035110674798488617
-8 0.0004034485900774598
-9 -0.0004926763358525932
-10 7.676315726712346e-05
+1 -0.0013786812778562307
```
Comment (liangel-02) [inline tests/assets/losses/rl_grpo_cuda.txt]: losses need to be updated bc we use _sum_dim_batch_invariant now
